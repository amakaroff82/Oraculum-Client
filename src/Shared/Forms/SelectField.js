import React from 'react';
import PropTypes from 'prop-types';
import { InputLabel } from 'material-ui/Input';
import { FormControl, FormHelperText } from 'material-ui/Form';
import SelectMenu from '../SelectMenu';
import { withStyles } from 'material-ui/styles';

// SelectField will show 4.5 menu items by default. Scrollbar will auto-show
// if there are more items. Use MenuProps to override this default.
const VISIBLE_MENU_ITEM_COUNT = 4.5;

// Hardcoded values from material-ui source
const MUI_MENU_ITEM_HEIGHT = 48;
const MUI_MENU_PADDING_TOP = 8;

const defaultMaxHeight =
  MUI_MENU_ITEM_HEIGHT * VISIBLE_MENU_ITEM_COUNT + MUI_MENU_PADDING_TOP;

const styles = theme => ({
  selectInput: {
    margin: '20px 0px',
  },
});

export const SelectField = ({
  className,
  classes,
  disabled,
  FormHelperTextProps,
  helperText,
  helperTextClassName,
  InputLabelProps,
  items,
  label,
  labelClassName,
  MenuProps,
  multiple,
  required,
  // redux-form props
  input,
  meta: { touched, error },
}) => {
  const hasError = touched && !!error;
  const hasHelperText = !!helperText;
  const separator = hasError && hasHelperText ? ' | ' : '';

  const menuProps = MenuProps || {
    PaperProps: { style: { maxHeight: defaultMaxHeight } },
  };

  return (
    <FormControl
      className={(className, classes.selectInput)}
      disabled={disabled}
      error={touched && !!error}
      fullWidth
      margin="normal"
      required={required}
      {...input}
    >
      {label && (
        <InputLabel
          htmlFor={input.name}
          className={labelClassName}
          {...InputLabelProps}
        >
          {label}
        </InputLabel>
      )}

      <SelectMenu
        InputId={input.name}
        InputProps={input}
        items={items}
        MenuProps={menuProps}
        multiple={multiple}
        onChange={value => input.onChange(value)}
        value={input.value}
      />

      {(hasError || hasHelperText) && (
        <FormHelperText
          className={helperTextClassName}
          {...FormHelperTextProps}
        >
          {hasError && error}
          {separator}
          {hasHelperText && helperText}
        </FormHelperText>
      )}
    </FormControl>
  );
};

SelectField.propTypes = {
  items: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      label: PropTypes.string,
    })
  ).isRequired,
};

export default withStyles(styles, { name: 'SelectField' })(SelectField);
