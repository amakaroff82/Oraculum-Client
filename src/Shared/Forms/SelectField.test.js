import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import SelectField from './SelectField';

const items = [
  { value: 'value1', label: 'label1' },
  { value: 'value2', label: 'label2' },
  { value: 'value3', label: 'label3' },
];
const value = items[1].value;

const defaultProps = {
  items,
  input: { value },
  meta: {},
};

const setupShallow = props => {
  props = { ...defaultProps, ...props };
  return shallow(<SelectField {...props} />);
};

describe('<SelectField /> component', () => {
  it('renders with default props correctly', () => {
    const selectField = setupShallow();
    expect(toJson(selectField)).toMatchSnapshot();
  });

  it('renders with label correctly', () => {
    const selectField = setupShallow({ label: 'foo' });
    expect(toJson(selectField)).toMatchSnapshot();
  });

  it('renders with helperText correctly', () => {
    const selectField = setupShallow({ helperText: 'bar' });
    expect(toJson(selectField)).toMatchSnapshot();
  });

  it('renders error if touched, concat with helperText if present', () => {
    const selectField = setupShallow({
      helperText: 'bar',
      meta: { touched: true, error: 'sad' },
    });
    expect(toJson(selectField)).toMatchSnapshot();
  });

  it('does not render error if not touched', () => {
    const selectField = setupShallow({
      meta: { touched: false, error: 'should not show error' },
    });
    expect(toJson(selectField)).toMatchSnapshot();
  });

  it('respects custom MenuProps', () => {
    const selectField = setupShallow({
      MenuProps: { elevation: 4 },
    });
    expect(toJson(selectField)).toMatchSnapshot();
  });
});
