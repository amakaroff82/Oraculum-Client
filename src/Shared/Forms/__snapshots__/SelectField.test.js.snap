// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectField /> component does not render error if not touched 1`] = `
<Component
  classes={
    Object {
      "selectInput": "SelectField-selectInput-1",
    }
  }
  input={
    Object {
      "value": "value2",
    }
  }
  items={
    Array [
      Object {
        "label": "label1",
        "value": "value1",
      },
      Object {
        "label": "label2",
        "value": "value2",
      },
      Object {
        "label": "label3",
        "value": "value3",
      },
    ]
  }
  meta={
    Object {
      "error": "should not show error",
      "touched": false,
    }
  }
/>
`;

exports[`<SelectField /> component renders error if touched, concat with helperText if present 1`] = `
<Component
  classes={
    Object {
      "selectInput": "SelectField-selectInput-1",
    }
  }
  helperText="bar"
  input={
    Object {
      "value": "value2",
    }
  }
  items={
    Array [
      Object {
        "label": "label1",
        "value": "value1",
      },
      Object {
        "label": "label2",
        "value": "value2",
      },
      Object {
        "label": "label3",
        "value": "value3",
      },
    ]
  }
  meta={
    Object {
      "error": "sad",
      "touched": true,
    }
  }
/>
`;

exports[`<SelectField /> component renders with default props correctly 1`] = `
<Component
  classes={
    Object {
      "selectInput": "SelectField-selectInput-1",
    }
  }
  input={
    Object {
      "value": "value2",
    }
  }
  items={
    Array [
      Object {
        "label": "label1",
        "value": "value1",
      },
      Object {
        "label": "label2",
        "value": "value2",
      },
      Object {
        "label": "label3",
        "value": "value3",
      },
    ]
  }
  meta={Object {}}
/>
`;

exports[`<SelectField /> component renders with helperText correctly 1`] = `
<Component
  classes={
    Object {
      "selectInput": "SelectField-selectInput-1",
    }
  }
  helperText="bar"
  input={
    Object {
      "value": "value2",
    }
  }
  items={
    Array [
      Object {
        "label": "label1",
        "value": "value1",
      },
      Object {
        "label": "label2",
        "value": "value2",
      },
      Object {
        "label": "label3",
        "value": "value3",
      },
    ]
  }
  meta={Object {}}
/>
`;

exports[`<SelectField /> component renders with label correctly 1`] = `
<Component
  classes={
    Object {
      "selectInput": "SelectField-selectInput-1",
    }
  }
  input={
    Object {
      "value": "value2",
    }
  }
  items={
    Array [
      Object {
        "label": "label1",
        "value": "value1",
      },
      Object {
        "label": "label2",
        "value": "value2",
      },
      Object {
        "label": "label3",
        "value": "value3",
      },
    ]
  }
  label="foo"
  meta={Object {}}
/>
`;

exports[`<SelectField /> component respects custom MenuProps 1`] = `
<Component
  MenuProps={
    Object {
      "elevation": 4,
    }
  }
  classes={
    Object {
      "selectInput": "SelectField-selectInput-1",
    }
  }
  input={
    Object {
      "value": "value2",
    }
  }
  items={
    Array [
      Object {
        "label": "label1",
        "value": "value1",
      },
      Object {
        "label": "label2",
        "value": "value2",
      },
      Object {
        "label": "label3",
        "value": "value3",
      },
    ]
  }
  meta={Object {}}
/>
`;
